### PR TITLE
Add CoverageChecksPublisher

### PR DESCRIPTION
**Pull request by XiongKezhi**
Example: https://github.com/XiongKezhi/codingstyle/pull/3/checks?check_run_id=888545163
- [x] Publish coverage as checks
- [x] Add tests
- [x] Add health report
- [x] Add configurations
- [ ] ...

### CODE REVIEW

**Comment by XiongKezhi** [comment]
cc @timja @uhafner @jeffpearce

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    private ChecksDetails extractChecksDetails() {
+        CoverageResult result = action.getResult();
+        ChecksOutputBuilder outputBuilder = new ChecksOutputBuilder()
+                .withTitle("Code Coverage")
+                .withSummary("## Generated " + result.getResults().size() + " types of coverage.")
+                .withText(extractChecksText(result)); // TODO: improve summary
+
+        return new ChecksDetailsBuilder()
+                .withName("Code Coverage")
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:36]: not sure which field it is as they have the same text currently but can a summary go on the check status on the right hand side

image

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        Map<CoverageElement, Ratio> ratios = result.getResults();
+        Map<String, Float> lastRatios = getLastRatios(result);
+
+        StringBuilder text = new StringBuilder();
+        for (Map.Entry<CoverageElement, Ratio> singleRatio : ratios.entrySet()) {
+            text.append("## ")
+                    .append(singleRatio.getKey().getName())
+                    .append(": \n* Coverage: ")
+                    .append(singleRatio.getValue())
+                    .append(" :white_check_mark:");
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:54]: consider putting this at the start of the line so that the emoji's are all aligned

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    private String extractChecksText(CoverageResult result) {
+        Map<CoverageElement, Ratio> ratios = result.getResults();
+        Map<String, Float> lastRatios = getLastRatios(result);
+
+        StringBuilder text = new StringBuilder();
+        for (Map.Entry<CoverageElement, Ratio> singleRatio : ratios.entrySet()) {
+            text.append("## ")
+                    .append(singleRatio.getKey().getName())
+                    .append(": \n* Coverage: ")
+                    .append(singleRatio.getValue())
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:53]: coverage is normally summarised as a percentage, not out of the number of lines covered

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    void publishChecks() {
+        ChecksPublisher publisher = ChecksPublisherFactory.fromRun(action.getOwner());
+        publisher.publish(extractChecksDetails());
+    }
+
+    private ChecksDetails extractChecksDetails() {
+        CoverageResult result = action.getResult();
+        ChecksOutputBuilder outputBuilder = new ChecksOutputBuilder()
+                .withTitle("Code Coverage")
+                .withSummary("## Generated " + result.getResults().size() + " types of coverage.")
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:32]: seems irrelevant to the user

File: pom.xml [hunk 5]
```diff
@@ -36,7 +39,7 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>structs</artifactId>
-            <version>1.9</version>
+            <version>1.20</version>
```
Comment (timja) [inline pom.xml:42]: use the jenkins bom instead of choosing all these versions directly

https://github.com/jenkinsci/bom

you could introduce it in a separate PR to reduce the diff here as well

File: pom.xml [hunk 6]
```diff
@@ -153,6 +171,10 @@
             <id>repo.jenkins-ci.org</id>
             <url>https://repo.jenkins-ci.org/public/</url>
         </repository>
+        <repository>
```
Comment (timja) [inline pom.xml:174]: shouldn't be needed,

run:
`mvn incrementals:incrementalify`

https://github.com/jenkinsci/incrementals-tools

File: pom.xml [hunk 7]
```diff
@@ -4,7 +4,7 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>3.9</version>
+        <version>4.0</version>
```
Comment (timja) [inline pom.xml:7]:
```suggestion
        <version>4.3</version>
```

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java (hunk 1 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:36]: `title` field in the `output` object. Then I'll change that part in warnings checks as well.

Comment (timja) [reply src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:36]: take a look at this plugin: https://github.com/jenkinsci/github-coverage-reporter-plugin

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+ */
+class CoverageChecksPublisher {
+    private final CoverageAction action;
+
+    CoverageChecksPublisher(final CoverageAction action) {
+        this.action = action;
+    }
+
+    void publishChecks() {
+        ChecksPublisher publisher = ChecksPublisherFactory.fromRun(action.getOwner());
```
Comment (jeffpearce) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:24]: Can this ever return null? If so, you don't want to do line 25

Comment (jeffpearce) [reply src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:24]: Oh, I see. It returns a `NullChecksPublisher` if it can't find one

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                        .append("%");
+            }
+
+            text.append("\n");
+        }
+
+        return text.toString();
+    }
+
+    private String extractChecksTitle(final CoverageResult result) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:81]: I would prefer to see line and branch coverage in the title.

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                .withStatus(ChecksStatus.COMPLETED)
+                .withConclusion(ChecksConclusion.SUCCESS)
+                .withDetailsURL(action.getAbsoluteUrl())
+                .withOutput(output)
+                .build();
+    }
+
+    private String extractChecksText(final CoverageResult result) {
+        Map<CoverageElement, Ratio> ratios = result.getResults();
+        Map<CoverageElement, Ratio> lastRatios = getLastRatios(result);
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:48]: I wonder if it would make sense to separate the total report and the delta report? Maybe we can discuss that topic in our meeting today. This question is also relevant for the analysis report (new and total warnings).

File: src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java [hunk 11]
```diff
@@ -92,12 +92,18 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
...
+        CoverageAction action = null;
         try {
-            processor.performCoverageReport(reportAdapters, reportDetectors, globalThresholds);
+            action = processor.performCoverageReport(reportAdapters, reportDetectors, globalThresholds);
         } catch (CoverageException e) {
             listener.getLogger().println(ExceptionUtils.getFullStackTrace(e));
             run.setResult(Result.FAILURE);
         }
+
+        if (action != null) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java:103]: Move inside the `try` then `action = null` and the `if` is not required at all.

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java [hunk 12]
```diff
@@ -89,14 +89,15 @@ public CoverageProcessor(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @N
...
      */
-    public void performCoverageReport(List<CoverageReportAdapter> reportAdapters, List<ReportDetector> reportDetectors, List<Threshold> globalThresholds)
+    public CoverageAction performCoverageReport(List<CoverageReportAdapter> reportAdapters, List<ReportDetector> reportDetectors, List<Threshold> globalThresholds)
             throws IOException, InterruptedException, CoverageException {
         Map<CoverageReportAdapter, List<CoverageResult>> results = convertToResults(reportAdapters, reportDetectors);
 
         CoverageResult coverageReport = aggregateReports(results);
         if (coverageReport == null) {
-            return;
+            return null;
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java:100]: This looks somewhat strange (but this is from the existing code). When does the report 
- return `null`
- throw an `CoverageException`

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+        StringBuilder text = new StringBuilder();
+        for (Map.Entry<CoverageElement, Ratio> singleRatio : ratios.entrySet()) {
+            text.append("## ")
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:52]: Wouldn't it be simpler to have a table with two rows (titles and values) that contains a column for each type? Then it would not occupy so much place.

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                .withOutput(output)
+                .build();
+    }
+
+    private String extractChecksText(final CoverageResult result) {
+        Map<CoverageElement, Ratio> ratios = result.getResults();
+        Map<CoverageElement, Ratio> lastRatios = getLastRatios(result);
+
+        StringBuilder text = new StringBuilder();
+        for (Map.Entry<CoverageElement, Ratio> singleRatio : ratios.entrySet()) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:51]: What is the default sorting? It looks ok in the PR results but I'm not sure if this is just some coincidence.

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        ChecksOutput output = new ChecksOutputBuilder()
+                .withTitle(extractChecksTitle(result))
+                .withSummary("")
+                .withText(extractChecksText(result))
+                .build();
+
+        return new ChecksDetailsBuilder()
+                .withName("Code Coverage")
+                .withStatus(ChecksStatus.COMPLETED)
+                .withConclusion(ChecksConclusion.SUCCESS)
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:40]: The coverage report also has a quality gate that we should check.

File: src/test/java/io/jenkins/plugins/coverage/CoverageChecksPublisherTest.java [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+        ChecksDetails expectedDetails = new ChecksDetailsBuilder()
+                .withName("Code Coverage")
+                .withStatus(ChecksStatus.COMPLETED)
+                .withConclusion(ChecksConclusion.SUCCESS)
+                .withDetailsURL("http://127.0.0.1:8080/job/pipeline-coding-style/job/PR-3/49/coverage")
+                .withOutput(new ChecksOutputBuilder()
+                        .withTitle("Line coverage of 60%.")
+                        .withSummary("")
+                        .withText("## Conditional\n* :white_check_mark: Coverage: 40%\n"
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/coverage/CoverageChecksPublisherTest.java:35]: (*Conditional* actually is the wrong name, but this is something that needs to be changed in the coverage API and not in your PR)

File: pom.xml (hunk 6 above)
Comment (XiongKezhi) [reply pom.xml:186]: but our plugin is currently not released, so we need this to download the checks dependencies

Comment (timja) [reply pom.xml:186]: That’s automatically added by the maven profile that incrementals adds behind the scenes

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java (hunk 9 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:100]: Done, see: https://github.com/XiongKezhi/codingstyle/pull/3/checks?check_run_id=907434646

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java (hunk 14 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java]: I refactored my code, see field `COVERAGE_TYPES` in `CoverageChecksPublisher`, so I think the default sorting order is no longer a problem for us.

File: pom.xml [hunk 17]
```diff
@@ -104,7 +107,34 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>jackson2-api</artifactId>
-            <version>2.8.11.3</version>
+            <version>2.11.1</version>
+        </dependency>
+        <dependency>
+            <groupId>io.jenkins.plugins</groupId>
+            <artifactId>checks-api</artifactId>
+            <version>0.1.0-rc4.fc43ef9447c7</version>
```
Comment (timja) [inline pom.xml:115]: release this?

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java (hunk 13 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java]: the table format: https://github.com/XiongKezhi/codingstyle/pull/3/checks?check_run_id=907434646

File: src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java (hunk 11 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java:103]: I just tried to keep the `try-catch` section as small as possible for easier exception locating, so isn't this a concern?

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java (hunk 12 above)
Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: I refactored my code to get `CoverageAction` directly from `run` instead of changing this `CoverageProcessor`.

**Comment by XiongKezhi** [comment]
Now I've used the table format for coverage and trend, added health report and failed message, example: https://github.com/XiongKezhi/codingstyle/pull/3/checks?check_run_id=909846128

The last thing to do is to enable configuration.

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+        }
+
+        return lineTitle + " " + branchTitle;
+    }
+
+    private String extractChecksTitle(final String elementName, final String targetBuildName,
+                                      final float coverage, final float coverageDiff) {
+        StringBuilder title = new StringBuilder()
+                .append(elementName)
+                .append(" coverage: ")
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:126]: I would remove this line, it's obvious you're talking about coverage and github is truncating it in the UI
```suggestion
```

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+                + trends;
+    }
+
+    private String extractChecksTitle(final CoverageResult result) {
+        Map<CoverageElement, Ratio> lastRatios = getLastRatios(result);
+
+        String lineTitle;
+        float lineCoverage = result.getCoverage(CoverageElement.LINE).getPercentageFloat();
+        if (result.getLinkToBuildThatWasUsedForComparison() != null) {
+            lineTitle = extractChecksTitle("Line", "target branch build", lineCoverage,
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:101]: more trimming
```suggestion
            lineTitle = extractChecksTitle("Line", "target branch", lineCoverage,
```

File: src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java [hunk 20]
```diff
@@ -54,6 +54,8 @@
     private boolean calculateDiffForChangeRequests = false;
     private boolean failBuildIfCoverageDecreasedInChangeRequest = false;
 
+    private boolean publishChecks = false;
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java:57]: Don’t we want to default to publishing? Change to skipPublishingChecks?

Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java:57]: either is ok, I'll change it

File: pom.xml (hunk 17 above)
Comment (XiongKezhi) [reply pom.xml:115]: I'm ready to release it now, but we may need to refactor some part since we have to publish checks for  `Queue.WaitingItem` instead of `Run` for `queued` state, is that ok? (I believe nothing changes in the consumer side after the that)

File: src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java [hunk 21]
```diff
@@ -197,12 +197,12 @@ public boolean getFailBuildIfCoverageDecreasedInChangeRequest() {
     }
 
     @DataBoundSetter
-    public void setPublishChecks(boolean publishChecks) {
-        this.publishChecks = publishChecks;
+    public void setSkipPublishingChecks(boolean skipPublishingChecks) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java:200]: Would it then make sense to use the same name in the warnings plugin as well? This will help users to write their pipeline scripts.

**Comment by XiongKezhi** [comment]
@jeffpearce @cizezsy We have finished most of the work, do you have time to review it? :)

**Comment by timja** [comment]
cc @tempora-mutantur

File: pom.xml (hunk 6 above)
Comment (timja) [reply pom.xml:186]: can you fix this please?

Comment (XiongKezhi) [reply pom.xml]: ok, done now.

File: src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java [hunk 22]
```diff
@@ -0,0 +1,199 @@
+package io.jenkins.plugins.coverage;
+
+import edu.hm.hafner.util.VisibleForTesting;
+import hudson.model.Run;
+import io.jenkins.plugins.checks.api.*;
+import io.jenkins.plugins.checks.api.ChecksOutput.ChecksOutputBuilder;
+import io.jenkins.plugins.checks.api.ChecksDetails.ChecksDetailsBuilder;
+import io.jenkins.plugins.coverage.targets.*;
+import io.jenkins.plugins.util.JenkinsFacade;
+import jnr.ffi.Struct;
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java:10]: what's this for?

Comment (XiongKezhi) [reply src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java]: oops, unused import...
